rewrite-java/src/main/java/org/openrewrite/java/ChangePackage.java

```diff
@@ -112,15 +114,48 @@ public TreeVisitor<?, ExecutionContext> getVisitor() {
                         }
                     }
                     stopAfterPreVisit();
+                } else if (tree instanceof SourceFileWithReferences) {
+                    SourceFileWithReferences cu = (SourceFileWithReferences) tree;
+                    boolean recursive = Boolean.TRUE.equals(ChangePackage.this.recursive);
+                    String recursivePackageNamePrefix = oldPackageName + ".";
+                    for (Reference ref : cu.getReferences().getReferences()) {
+                        if (ref.getValue().equals(oldPackageName) || recursive && ref.getValue().startsWith(recursivePackageNamePrefix)) {
+                            return SearchResult.found(cu);
+                        }
+                    }
+                    stopAfterPreVisit();
                 }
                 return super.preVisit(tree, ctx);
             }
         };
 
-        return Preconditions.check(condition, new ChangePackageVisitor());
+        return Preconditions.check(condition, new TreeVisitor<Tree, ExecutionContext>() {
+            @Override
+            public boolean isAcceptable(SourceFile sourceFile, ExecutionContext ctx) {
+                return sourceFile instanceof JavaSourceFile || sourceFile instanceof SourceFileWithReferences;
+            }
+
+            @Override
+            public @Nullable Tree preVisit(@Nullable Tree tree, ExecutionContext ctx) {
```

Comment:
Also here we should make sure to call stopAfterPreVisit() once we have found everything we are after. Otherwise the visitor will continue to traverse the entire tree.